tests/pytorch/test_fused_rope.py

```diff
@@ -41,20 +42,118 @@ def apply_rotary_pos_emb_thd(
         cu_seqlens(Tensor):  Cumulative sum of sequence lengths in a batch for `t`,
         with shape [b + 1] and dtype torch.int32.
         freqs (Tensor): Rotary Positional embedding tensor freq is of shape [max_s, 1, 1, d]
+        start_positions (Tensor): Tensor of shape [b] determining the beginning offsets
+                         of frequeuncies applied to  sequences.
 
     Returns:
         Tensor: Shape [t, h, d]. The input tensor after applying RoPE.
     """
     cu_seqlens = cu_seqlens // cp_size
     seqlens = (cu_seqlens[1:] - cu_seqlens[:-1]).tolist()
-    return torch.cat(
-        [
-            apply_rotary_pos_emb(
-                x.unsqueeze(1), _get_thd_freqs_on_this_cp_rank(cp_rank, cp_size, x, freqs)
-            )
-            for x in torch.split(t, seqlens)
-        ]
-    ).squeeze(1)
+    if cp_size > 1:
+        return torch.cat(
+            [
+                apply_rotary_pos_emb(
+                    x.unsqueeze(1), _get_thd_freqs_on_this_cp_rank(cp_rank, cp_size, x, freqs)
+                )
+                for x in torch.split(t, seqlens)
+            ]
+        ).squeeze(1)
+    else:
```

Comment:
I feel it's better to change the if branches to
```python
if start_positions is None:
    assert cp_size == 1
    ...
else:
    # keep the original code
```
BTW, does `start_positions` works with CP? If so, we should enable it for the CP path as well. With #1626, the fused and unfused versions have the same support matrix for qkv formats/cp, etc.

Reply:
`start_positions` argument is used for inference use-cases where the input tokens after the "prefill" phase is over, have positions starting at an offset from left. And inference as far as we know doesn't use CP. But the `start_positions` argument could theoretically be used to offset the RoPE embeddings for all the CP ranks and it should work fine.

I'll add some doc/note in the code that makes this explicit